mne/beamformer/_rap_music.py

```diff
@@ -0,0 +1,239 @@
+"""Compute Compute a Recursively Applied and Projected MUltiple
+Signal Classification (RAP-MUSIC).
+"""
+
+# Authors: Yousra Bekhti <yousra.bekhti@gmail.com>
+#
+# License: BSD (3-clause)
+
+import warnings
+
+import numpy as np
+from scipy import linalg
+
+from ..io.constants import FIFF
+from ..io.pick import pick_channels_evoked
+from ..forward import _subject_from_forward
+from ..minimum_norm.inverse import _check_reference
+from ..cov import compute_whitener
+from ..source_estimate import _make_stc
+from ..utils import logger, verbose
+from ._lcmv import _prepare_beamformer_input
+
+
+@verbose
+def _apply_rap_music(data, info, tmin, forward, noise_cov, label=None,
+                     r=15, n_sources=5, picks=None, pick_ori=None,
+                     return_residual=False, verbose=None):
+    """ RAP-MUSIC for evoked data
+
+    Parameters
+    ----------
+    data : array or list / iterable
+        Evoked data.
+    info : dict
+        Measurement info.
+    tmin : float
+        Time of first sample.
+    forward : dict
+        Forward operator.
+    noise_cov : Covariance
+        The noise covariance.
+    label : Label
+        Restricts the rap-music solution to a given label.
+        XXX: not implemented yet.
+    r: int
+        The dimension of the subspace spanning the signal.
+        The default value is 15.
+    n_sources: int
+        The number of sources to estimate.
+    picks : array-like of int | None
+        Indices (in info) of data channels. If None, MEG and EEG data channels
+        (without bad channels) will be used.
+    pick_ori : None | 'normal' | 'max-power'
+        If 'normal', rather than pooling the orientations by taking the norm,
+        only the radial component is kept. If 'max-power', the source
+        orientation that maximizes output source power is chosen.
+        XXX: 'max-power' not implemented
+    return_residual : bool
+        If True, the residual is returned as an Evoked instance.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+
+    Returns
+    -------
+    stc : SourceEstimate
+        Source time courses.
+    D : array
+        Data explained by the sources. Computed only if return_residual
+        is True.
+    """
+    is_free_ori, picks, ch_names, proj, vertno, G =\
+        _prepare_beamformer_input(info, forward, label, picks, pick_ori)
+
+    # Handle whitening + data covariance
+    whitener, _ = compute_whitener(noise_cov, info, picks)
+
+    # whiten the leadfield
+    G = np.dot(whitener, G)
+    data = np.dot(whitener, data)
+
+    # Pick source orientation normal to cortical surface
+    if pick_ori == 'normal':
+        G = G[:, 2::3]
+        is_free_ori = False
+
+    eig_values, eig_vectors = linalg.eigh(np.dot(data, data.T))
+    phi_sig = eig_vectors[:, -r:]
```

Comment:
> On Feb 21, 2015, at 01:20, Alexandre Gramfort <notifications@github.com> wrote:
> 
> In mne/beamformer/_rap_music.py:
> 
> > +        _prepare_beamformer_input(info, forward, label, picks, pick_ori)
> > +
> > +    # Handle whitening + data covariance
> > +    whitener, _ = compute_whitener(noise_cov, info, picks)
> > +
> > +    # whiten the leadfield
> > +    G = np.dot(whitener, G)
> > +    data = np.dot(whitener, data)
> > +
> > +    # Pick source orientation normal to cortical surface
> > +    if pick_ori == 'normal':
> > +        G = G[:, 2::3]
> > +        is_free_ori = False
> > +
> > +    eig_values, eig_vectors = linalg.eigh(np.dot(data, data.T))
> > +    phi_sig = eig_vectors[:, -r:]
> I would have guessed eig_vectors[:, :r] no?
> 
Yes actually me too. But the eig_values were in the other order (ascending). Will check it again.

> —
> Reply to this email directly or view it on GitHub.
>